File: src/clipboard.ts
import { mkdtemp } from "node:fs/promises";
import { tmpdir } from "node:os";
import { dirname, join, resolve } from "node:path";
import { fileURLToPath } from "node:url";
import { randomUUID } from "node:crypto";
import { runLocal } from "./ssh.js";

export interface CapturedImage {
  path: string;
  mimeType: string;
  sizeBytes: number;
}

export async function captureClipboardImage(): Promise<CapturedImage> {
  const dir = await mkdtemp(join(tmpdir(), "ssh-bin-paste-"));
  const outputPath = join(dir, `${randomUUID()}.png`);
  const script = await clipboardHelperPath();
  const result = await runLocal("swift", [script, "--output", outputPath]);
  if (result.exitCode !== 0) {
    throw new Error(result.stderr.trim() || "No image found on the clipboard");
  }

  try {
    return JSON.parse(result.stdout) as CapturedImage;
  } catch {
    throw new Error(`Clipboard helper returned invalid JSON: ${result.stdout}`);
  }
}

async function clipboardHelperPath(): Promise<string> {
  const here = dirname(fileURLToPath(import.meta.url));
  const candidates = [
    resolve(here, "../native/clipboard-image.swift"),
    resolve(here, "../../native/clipboard-image.swift"),
  ];
  for (const candidate of candidates) {
    return candidate;
  }
  throw new Error("Could not find clipboard helper");
}

File: src/clipboard.rs
use anyhow::{Context, Result, bail};
use serde::Deserialize;
use uuid::Uuid;

use crate::assets::{CLIPBOARD_IMAGE_SWIFT, write_temp_asset};
use crate::ssh::run_local;

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CapturedImage {
    pub path: String,
    pub mime_type: String,
    pub size_bytes: u64,
}

pub fn capture_clipboard_image() -> Result<CapturedImage> {
    let script = write_temp_asset("clipboard-image.swift", CLIPBOARD_IMAGE_SWIFT)?;
    let output_path = std::env::temp_dir()
        .join(format!("ssh-bin-paste-{}.png", Uuid::new_v4()))
        .display()
        .to_string();
    let result = run_local(
        "swift",
        &[script.to_str().unwrap_or(""), "--output", &output_path],
        None,
    )?;
    if result.exit_code != 0 {
        bail!(
            "{}",
            result
                .stderr
                .trim()
                .strip_suffix('\n')
                .unwrap_or(result.stderr.trim())
        );
    }
    serde_json::from_str(&result.stdout).context("clipboard helper returned invalid JSON")
}

